src/aapns/connection.py

```diff
@@ -0,0 +1,440 @@
+from __future__ import annotations
+
+import json
+from asyncio import (
+    CancelledError,
+    Event,
+    Future,
+    TimeoutError,
+    create_task,
+    open_connection,
+    wait_for,
+)
+from contextlib import suppress
+from dataclasses import dataclass, field
+from logging import getLogger
+from math import inf
+from ssl import OP_NO_TLSv1, OP_NO_TLSv1_1, SSLContext, SSLError, create_default_context
+from time import time
+from typing import Dict, List, Optional
+from urllib.parse import urlparse
+
+import h2.config
+import h2.connection
+import h2.settings
+
+from .errors import Blocked, Closed, FormatError, ResponseTooLarge, Timeout
+
+# Apple limits APN payload (data) to 4KB or 5KB, depending.
+# Request header is not subject to flow control in HTTP/2
+# Data is subject to framing and padding, but those are minor.
+MAX_NOTIFICATION_PAYLOAD_SIZE = 5120
+REQUIRED_FREE_SPACE = 6000
+# OK response is empty
+# Error response is short json, ~30 bytes in size
+MAX_RESPONSE_SIZE = 2 ** 16
+# Inbound connection flow control window
+# It's quite arbitrary, guided by:
+# * concurrent requests limit, server limit being 1000 today
+# * expected response size, see above
+CONNECTION_WINDOW_SIZE = 2 ** 24
+logger = getLogger(__package__)
+
+
+@dataclass(eq=False)
+class Connection:
+    """Encapsulates a single HTTP/2 connection to the APN server
+
+    Connection states:
+    * new (not connected)
+    * starting
+    * active
+    * graceful shutdown (to do)
+    * closing
+    * closed
+    """
+
+    host: str
+    port: int
+    protocol: h2.connection.H2Connection
+    read_stream: Any
+    write_stream: Any
+    should_write: Event
+    channels: Dict[int, Channel] = field(default_factory=dict)
+    reader: Task = field(init=False)
+    writer: Task = field(init=False)
+    closed: bool = False
+    closing: bool = False
+    outcome: Optional[str] = None
+    max_concurrent_streams: int = 100  # recommended in RFC7540#section-6.5.2
+    last_new_sid: int = -1
+    last_sent_sid: int = -1  # client streams are odd
+
+    def __repr__(self):
+        bits = [self.state, f"{self.host}:{self.port}"]
+        if self.state != "closed":
+            bits.extend([f"buffered:{self.buffered}", f"inflight:{self.inflight}"])
+        else:
+            bits.append(f"outcome:{self.outcome}")
+        return "<Connection %s>" % " ".join(bits)
```

Comment:
custom repr no longer needed with dataclass? (you can omit things with `field(repr=False)`)

Reply:
Ish... but then I'd need to export computed attributes in some other way.
Like in-flight request count.
Maybe that's better anyway 🤔

Reply:
forgot `buffered` and `inflight` are computed. maybe there should be a `statistics` computed property that returns an dataclass with the statistics (so you'd repr(conn.statistics) rather than repr(conn) if you want that info).

Reply:
Good idea!
(I have a nagging feeling that none will actually use it though, it's hidden so many layers deep)

Reply:
ideally nobody will *need* to use it because it'll work flawless 😂